helix-core/src/main/java/org/apache/helix/controller/GenericHelixController.java

```diff
@@ -695,12 +695,19 @@ private void handleEvent(ClusterEvent event, BaseControllerDataProvider dataProv
     event.addAttribute(AttributeName.STATEFUL_REBALANCER.name(),
         _rebalancerRef.getRebalancer(manager));
 
-    if (!manager.isLeader()) {
-      logger.error("Cluster manager: " + manager.getInstanceName() + " is not leader for " + manager
-          .getClusterName() + ". Pipeline will not be invoked");
+    Optional<String> leaderSession = manager.getSessionIdIfLead();
```

Comment:
@jiajunwang 
> can we actually add the session Id to the ClusterEvent when it is constructed, and then use that session Id directly here?

I don't think it is accurate: when the event is constructed and we add this session id **S0** to event, if the manager session id changes to **S1** when handling the event, the event would not be processed. But actually I think it would be more accurate to use the session id when starting to process the pipeline.

Or even session id changes to S1 and the event is not handled, there will be future events to keep the correctness. In this case, I think it is fine to add the session id to when event is created. But I think it may affect a bit performance: originally event0 will process the pipeline, but it is discarded and next event1 will be processed later.

So I think it is better to add the session right before running the pipeline.

Reply:
The reason we want this change is to prevent 2 pipelines running. So if the S0 session event is handled in S1, do you think it would be risky that we have more confusion and leads to a bad result?

Reply:
Synced with @jiajunwang offline. We agreed to add event session when event is created. Updated the PR.